### ISSUE TITLE
`jnp.sort` and `jnp.vdot` raise error under `jax.disable_jit()`

### ISSUE THREAD

**Issue by soraros** [google/jax#14906]
### Description

`jnp.sort`, `jnp.vdot` and a few others in `jax.numpy` will raise under `jax.disable_jit()`. IMO, the fact that exactly *these* functions go wrong is not important, nor is the fact that `jax.disable_jit()` is involved.

Taking a look at their implementation, on can spot a pattern:

```python
@partial(jit, static_argnums=(...))
def f(a: ArrayLike):
  util._check_arraylike('f', a)
  bar = foo(a.ravel())  # << not good
  ...
```

Without jit, the scalar case will get caught by the tests, lead us to the fix naturally. These functions do array conversion implicitly on the jit boundary hides the fact that array methods might not be available.

Fixing them is easy enough, but I want to ask/discuss something more general here:
1. Maybe we should insert these explicit conversions after `_check_arraylike` everywhere, in the form of `asarray` or explicit `reshape` or `ravel`.
2. If the pattern in 1. is desirable, is it always safe to use the simpler and faster variant `util._asarray`? I think if we can, we should.
3. I see similar problems (that JAX maintainers have rather canonical solutions in JAX itself) popping up almost every time someone tries to contribute to `jax.scipy` etc. Can we have a contributor guide documenting these issues? (`array` conversion, `dtype` handling, type hinting, `jnp` vs `lax`, standard ways to deal with static arguments: `concrete_or_error`, etc.)
4. What is the best place to ask these broader questions that are also half bug reports? Should the Github discussions be preferred?

@jakevdp I bore you the most in many related issues before, so I'm gonna tag you here.

**Comment by jakevdp** [comment]
Agreed we should always call `_check_arraylike` followed by casting the value using `asarray` in order to address this problem, though it's a common-enough pattern we might want to replace it with a more streamlined utility.

**Comment by jakevdp** [comment]
Yeah, that's roughly what I've been thinking. Though for typing, if you want to use array attributes you have to rename the variables, e.g.
```python
a_arr, b_arr = util._arraylike(a, b)
```

**Comment by soraros** [comment]
> It's one of my biggest annoyances with the current generation of static type checkers, actually

Yeah. Pyright seems to work without renaming though, so there is hope.

**Comment by jakevdp** [comment]
Fix for this particular issue here: #14908

I'll probably tackle the simplification of `check_arraylike` in a later PR – it's a bigger job.

**Comment by soraros** [comment]
@jakevdp Thanks!
